interpret: add allocation parameters to `AllocBytes`

nia-e opened the pull request:
Necessary for a better implementation of [rust-lang/miri#4343](https://github.com/rust-lang/miri/pull/4343). Also included here is the code from that PR, adapted to this new interface for the sake of example and so that CI can run on them; the Miri changes can be reverted and merged separately, though.

r? @RalfJung

RalfJung commented:
So you don't want a review yet? :thinking:

RalfJung commented on src/tools/miri/src/alloc/alloc_bytes.rs:
Something odd happened here and there's now two copies of `MiriAllocBytes` in the repo with this PR?^^

Ideally, remove the Miri changes from this PR, and have it contain just the parts needed for the new allocator parameter.

nia-e commented:
Feel free to go ahead, but it's not fully implemented. I realised right after opening this there was more to do, oopsie

nia-e committed:
extend allocbytes with associated type

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15231107448/job/42839071051) [(plain)](https://github.com/rust-lang/rust/commit/15a3f38ff3ec7f7e34997e6f1d71938674a7b923/checks/42839071051/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking miri v0.1.0 (/checkout/src/tools/miri)
error[E0050]: method `from_bytes` has 2 parameters but the declaration in trait `from_bytes` has 3
  --> src/tools/miri/src/alloc_bytes.rs:92:30
   |
92 |     fn from_bytes<'a>(slice: impl Into<Cow<'a, [u8]>>, align: Align) -> Self {
   |                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected 3 parameters, found 2
   |
   = note: `from_bytes` from trait: `fn(impl Into<Cow<'a, [u8]>>, Align, <Self as rustc_const_eval::interpret::AllocBytes>::ByteMetadata) -> Self`

error[E0050]: method `zeroed` has 2 parameters but the declaration in trait `rustc_const_eval::interpret::AllocBytes::zeroed` has 3
   --> src/tools/miri/src/alloc_bytes.rs:108:21
    |
108 |     fn zeroed(size: Size, align: Align) -> Option<Self> {
    |                     ^^^^^^^^^^^^^^^^^^ expected 3 parameters, found 2
    |
    = note: `zeroed` from trait: `fn(rustc_abi::Size, Align, <Self as rustc_const_eval::interpret::AllocBytes>::ByteMetadata) -> std::option::Option<Self>`

error[E0061]: this function takes 3 arguments but 2 arguments were supplied
   --> src/tools/miri/src/alloc_addresses/mod.rs:142:46
    |
142 |                         let prepared_bytes = MiriAllocBytes::zeroed(info.size, info.align)
    |                                              ^^^^^^^^^^^^^^^^^^^^^^----------------------- argument #3 of type `()` is missing
    |
note: associated function defined here
   --> /checkout/compiler/rustc_middle/src/mir/interpret/allocation.rs:43:8
    |
43  |     fn zeroed(size: Size, _align: Align, _dsc: Self::ByteMetadata) -> Option<Self>;
    |        ^^^^^^
help: provide the argument
    |
142 |                         let prepared_bytes = MiriAllocBytes::zeroed(info.size, info.align, ())
    |                                                                                          ++++

error[E0061]: this function takes 3 arguments but 2 arguments were supplied
   --> src/tools/miri/src/alloc_addresses/mod.rs:162:25
    |
162 |                         MiriAllocBytes::from_bytes(&[0u8; 1], Align::from_bytes(1).unwrap());
    |                         ^^^^^^^^^^^^^^^^^^^^^^^^^^------------------------------------------ argument #3 of type `()` is missing
    |
note: associated function defined here
   --> /checkout/compiler/rustc_middle/src/mir/interpret/allocation.rs:35:8
    |
35  |     fn from_bytes<'a>(
    |        ^^^^^^^^^^
help: provide the argument
    |
162 |                         MiriAllocBytes::from_bytes(&[0u8; 1], Align::from_bytes(1).unwrap(), ());
    |                                                                                            ++++

error[E0061]: this function takes 3 arguments but 2 arguments were supplied
   --> src/tools/miri/src/alloc_addresses/mod.rs:432:23
    |
432 |             interp_ok(MiriAllocBytes::from_bytes(std::borrow::Cow::Borrowed(bytes), align))
    |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^------------------------------------------ argument #3 of type `()` is missing
    |
note: associated function defined here
   --> /checkout/compiler/rustc_middle/src/mir/interpret/allocation.rs:35:8
    |
35  |     fn from_bytes<'a>(
    |        ^^^^^^^^^^
help: provide the argument
    |
432 |             interp_ok(MiriAllocBytes::from_bytes(std::borrow::Cow::Borrowed(bytes), align, ()))
    |                                                                                          ++++

error[E0061]: this function takes 3 arguments but 2 arguments were supplied
   --> src/tools/miri/src/concurrency/thread.rs:902:31
    |
902 |                     interp_ok(MiriAllocBytes::from_bytes(std::borrow::Cow::Borrowed(bytes), align))
    |                               ^^^^^^^^^^^^^^^^^^^^^^^^^^------------------------------------------ argument #3 of type `()` is missing
    |
note: associated function defined here
   --> /checkout/compiler/rustc_middle/src/mir/interpret/allocation.rs:35:8
    |
35  |     fn from_bytes<'a>(
    |        ^^^^^^^^^^
help: provide the argument
    |
902 |                     interp_ok(MiriAllocBytes::from_bytes(std::borrow::Cow::Borrowed(bytes), align, ()))
    |                                                                                                  ++++

Some errors have detailed explanations: E0050, E0061.
For more information about an error, try `rustc --explain E0050`.
[RUSTC-TIMING] miri test:false 3.410
```

</details>

RalfJung commented on compiler/rustc_middle/src/mir/interpret/allocation.rs:
```diff
@@ -27,12 +27,20 @@ use crate::ty;
+    fn from_bytes<'a>(
+        slice: impl Into<Cow<'a, [u8]>>,
+        _align: Align,
+        _dsc: Self::ByteMetadata,
```
Please pick a consistent name: is this metadata or is it a descriptor?

I would have used neither term and called in `AllocParams`.

nia-e replied:
Yeah ok I was all over the place with the names, initially I wanted to call it the discriminant but a similar name was already taken. I'll use yours though ^^

RalfJung commented on compiler/rustc_middle/src/mir/interpret/allocation.rs:
```diff
@@ -27,12 +27,20 @@ use crate::ty;
 pub trait AllocBytes: Clone + fmt::Debug + Deref<Target = [u8]> + DerefMut<Target = [u8]> {
+    /// Used for giving extra metadata on creation. Miri uses this to allow
+    /// machine memory to be allocated separately from other allocations.
+    type ByteMetadata: Clone + fmt::Debug;
```
Why is this `Clone`?

RalfJung commented on compiler/rustc_middle/src/mir/interpret/allocation.rs:
```diff
@@ -47,15 +55,20 @@ pub trait AllocBytes: Clone + fmt::Debug + Deref<Target = [u8]> + DerefMut<Targe
     fn as_ptr(&self) -> *const u8;
+
+    /// Gets the allocation metadata.
+    fn get_mdata(&self) -> Self::ByteMetadata;
```
What is this needed for?

nia-e replied:
It is not! I left it over from an earlier attempt at making this work, I'll delete it

RalfJung commented on compiler/rustc_middle/src/mir/interpret/allocation.rs:
```diff
@@ -175,6 +190,7 @@ fn all_zero(buf: &[u8]) -> bool {
 impl<Prov: Provenance, Extra, Bytes, E: Encoder> Encodable<E> for Allocation<Prov, Extra, Bytes>
```
Please just restrict this and the next instance to an allocation using `Box<[u8]>` for the bytes.

RalfJung reviewed (commented):
Thanks! However, I have some questions. :)

@rustbot author

RalfJung commented:
> Ok! So I spent an hour or two panicking over some failing tests only to realise that they're failing on master, actually

I see no test failure in CI here? If some tests fail on your machine, please file an issue with instructions to reproduce the problem.

nia-e commented:
Seems like the failing tests are just a me thing, unsure what it is but I'm not going to debug it if it's not a big deal (+ it's only an issue when doing `./x test miri` and it might just be fixed by doing a `cargo clean` but I don't feel like waiting two weeks for it ro recompile after). I pushed something that hopefully addresses the review!

@rustbot ready

RalfJung commented on src/tools/miri/src/machine.rs:
```diff
@@ -1804,6 +1804,9 @@ impl<'tcx> Machine<'tcx> for MiriMachine<'tcx> {
     }
+
+    /// Placeholder!
+    fn get_default_alloc_params(&self) -> () {}
```
```suggestion
    fn get_default_alloc_params(&self) -> MiriAllocBytes::AllocParams { () }
```

RalfJung reviewed (commented):
Looks good, just some minor comments. :)

@rustbot author

nia-e commented:
@rustbot ready
